### PR TITLE
Added disk space warning if not enough for ES reindex

### PR DESCRIPTION
**Pull request by jashaik**
Signed-off-by: jan shahid shaik <jashaik@progress.com>

### Description
Raise disk space warning if disk space is not enough for ES reindexing

### Issues Resolved
Added function to verify disk space in the preflight checks 

https://app.zenhub.com/workspaces/chef-infra-server-team-5fc64867d45ca500173dbbc7/issues/chef/chef-server/2197
### Check List

- [ ] New functionality includes tests
- [x] All buildkite tests pass
- [ ] Full omnibus build and tests in buildkite pass
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [ ] PR title is a worthy inclusion in the CHANGELOG

### CODE REVIEW

**Comment by jashaik** [comment]
Pipeline information 
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2227

**Comment by jashaik** [comment]
Implemented as per the _a2_ha_backend - https://chefio.slack.com/archives/CEJ96JKRT/p1614969716000200?thread_ts=1614855139.033500&cid=CEJ96JKRT

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb [hunk 1]
```diff
@@ -89,6 +90,24 @@ def verify_heap_size
     end
   end
 
+  def verify_disk_space
+    data_dir = node_elasticsearch_attr["data_dir"]
+    if Dir.exist?(data_dir)
+      data_dir_size = Du.du(data_dir)
+      free_disk_space = Statfs.new("#{data_dir}/..").free_space
+      if (2* data_dir_size) < free_disk_space
```
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb:98]: could you include a comment here about why we are choosing this number?

**Comment by lbakerchef** [comment]
Is there any good way to test this?  Maybe 'mocking' or 'faking' things somehow so that the disk space comes back as insufficient, the directory comes back missing, etc?

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb [hunk 2]
```diff
@@ -89,6 +90,27 @@ def verify_heap_size
     end
   end
 
+  # Disk space verification for running reindex 
+  # It needs enough to hold a second copy of the data plus whatever change in size the new index might need if the schema has changed.
+  # Required Disk Space = (2 * size of ES data directory)
+  def verify_disk_space
+    data_dir = node_elasticsearch_attr["data_dir"]
+    if Dir.exist?(data_dir)
+      data_dir_size = Du.du(data_dir)
```
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb:99]: nice!

**Comment by PrajaktaPurohit** [comment]
I wonder is https://github.com/chef/chef-server/blob/master/src/chef-server-ctl/plugins/reindex.rb is a better place to add the check than the preflight checks?

**Comment by jashaik** [comment]
> Is there any good way to test this? Maybe 'mocking' or 'faking' things somehow so that the disk space comes back as insufficient, the directory comes back missing, etc?

We can mock this by using load test tools but I am not having idea about our load testing tools. The directory missing can be seen in fresh installs which is handled by 'Dir.exist?'

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb [hunk 3]
```diff
@@ -89,6 +90,27 @@ def verify_heap_size
...
 
+  # Disk space verification for running reindex 
+  # It needs enough disk space to hold a second copy of the data plus whatever change in size the new index might need if the schema has changed.
+  # Required Disk Space = (2 * size of ES data directory)
+  def verify_disk_space
+    data_dir = node_elasticsearch_attr["data_dir"]
+    if Dir.exist?(data_dir)
+      data_dir_size = Du.du(data_dir)
+      free_disk_space = Statfs.new("#{data_dir}/..").free_space
+      if (2* data_dir_size) < free_disk_space #The minimum space should be double the existing data size
```
Comment (PrajaktaPurohit) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_indexing_validator.rb:101]: Might be useful to consider a multiplying factor of 2.1 or 2.2 here?

**Pull request closed without merging**